src/main.c

```diff
@@ -464,10 +466,21 @@ static void async_nb_read_ready(h2o_socket_t *sock, const char *err);
 static void async_nb_on_write_complete(h2o_socket_t *sock, const char *err)
 {
     /* add to the read queue the entry for which we have written the request, and start reading from the socket */
-    async_nb_push(&async_nb.read_queue, async_nb_pop(&async_nb.write_queue));
+    struct async_nb_transaction_t *transaction = async_nb_pop(&async_nb.write_queue);
+    if (transaction->expect_response) {
+        async_nb_push(&async_nb.read_queue, transaction);
+    }
+
     if (!h2o_socket_is_reading(async_nb.sock))
         h2o_socket_read_start(async_nb.sock, async_nb_read_ready);
```

Comment:
Maybe this block has to be part of the newly introduced `if (transaction->expect_response)`?

Rationale is that we want to read from the socket only when we expect a response.